File: js/kd-tree.js
/**
 * ============================================================
 *  KD-Tree (K-Dimensional Tree) - 2 Boyutlu Uzay Ağacı
 * ============================================================
 * 
 * Durakların 2 boyutlu koordinatlarını indeksler.
 * Verilen bir noktaya en yakın K durağı verimli şekilde bulur.
 * 
 * Çalışma Prensibi:
 *   - Her seviyede bir eksen (x veya y) seçilir
 *   - Düğüm, o eksendeki medyan noktadır
 *   - Sol alt ağaç: daha küçük koordinatlı noktalar
 *   - Sağ alt ağaç: daha büyük koordinatlı noktalar
 * 
 * Zaman Karmaşıklığı:
 *   - Ağaç Oluşturma: O(N log N)        (sıralama + özyineleme)
 *   - En Yakın Komşu:  O(log N) ortalama, O(N) en kötü
 *   - KNN Sorgusu:     O(K log N) ortalama
 *   - Ekleme:          O(log N) ortalama
 * 
 * Uzay Karmaşıklığı: O(N)
 * ============================================================
 */

/**
 * KD-Tree düğüm sınıfı
 * Her düğüm bir noktayı, derinliğini ve çocuklarını tutar
 */
class KdTreeNode {
    constructor(point, depth = 0) {
        this.point = point;       // { x, y, id, name, ... }
        this.depth = depth;       // Ağaçtaki derinlik (eksen seçimi için)
        this.left = null;         // Sol çocuk (daha küçük değerler)
        this.right = null;        // Sağ çocuk (daha büyük değerler)
    }
}

export class KdTree {
    /**
     * KD-Tree oluşturucusu
     * @param {Array} points - { x, y, id, name, ... } nesnelerinin dizisi
     */
    constructor(points = []) {
        this.dimensions = 2;      // 2D: x ve y eksenleri
        this.nodeCount = 0;       // Toplam düğüm sayısı
        
        // Ağacı dengeli olarak inşa et
        this.root = this._buildBalanced([...points], 0);
    }

    // --- Ağaç İnşası ---

    /**
     * Dengeli KD-Tree inşa eder
     * 
     * Strateji:
     *   1. Mevcut eksene göre noktaları sırala
     *   2. Ortadaki noktayı (medyan) kök yap
     *   3. Sol yarıyı sol alt ağaca, sağ yarıyı sağ alt ağaca ver
     * 
     * @param {Array} points - Nokta dizisi
     * @param {number} depth - Mevcut derinlik
     * @returns {KdTreeNode|null}
     * @complexity O(N log N)
     */
    _buildBalanced(points, depth) {
        if (points.length === 0) return null;

        // Hangi eksende böleceğimizi belirle (derinlik mod 2)
        const axis = depth % this.dimensions;
        const key = axis === 0 ? 'x' : 'y';

        // O eksende sırala
        points.sort((a, b) => a[key] - b[key]);

        // Medyanı seç
        const medianIndex = Math.floor(points.length / 2);
        
        const node = new KdTreeNode(points[medianIndex], depth);
        this.nodeCount++;

        // Özyinelemeli olarak alt ağaçları oluştur
        node.left = this._buildBalanced(
            points.slice(0, medianIndex), 
            depth + 1
        );
        node.right = this._buildBalanced(
            points.slice(medianIndex + 1), 
            depth + 1
        );

        return node;
    }

    // --- Ekleme ---

    /**
     * Ağaca yeni bir nokta ekler
     * @param {Object} point - { x, y, id, name, ... }
     * @complexity O(log N) ortalama
     */
    insert(point) {
        this.root = this._insertNode(this.root, point, 0);
    }

    _insertNode(node, point, depth) {
        if (node === null) {
            this.nodeCount++;
            return new KdTreeNode(point, depth);
        }

        const axis = depth % this.dimensions;
        const key = axis === 0 ? 'x' : 'y';

        if (point[key] < node.point[key]) {
            node.left = this._insertNode(node.left, point, depth + 1);
        } else {
            node.right = this._insertNode(node.right, point, depth + 1);
        }

        return node;
    }

    // --- En Yakın Komşu (Nearest Neighbor) ---

    /**
     * Verilen noktaya en yakın durağı bulur
     * 
     * Algoritma:
     *   1. Hedef noktaya uygun dala in
     *   2. Yaprak düğümden geri dönerken mesafeyi kontrol et
     *   3. Diğer dalı kontrol etmemiz gerekip gerekmediğine karar ver
     *      (bölme düzlemine olan mesafe < en iyi mesafe ise kontrol et)
     * 
     * @param {Object} target - { x, y } hedef nokta
     * @returns {Object|null} En yakın nokta
     * @complexity O(log N) ortalama, O(N) en kötü
     */
    nearestNeighbor(target) {
        if (this.root === null) return null;

        const best = { point: null, distSq: Infinity };
        this._searchNearest(this.root, target, best);
        
        return best.point ? {
            ...best.point,
            distance: Math.sqrt(best.distSq)
        } : null;
    }

    _searchNearest(node, target, best) {
        if (node === null) return;

        // Mevcut düğüme olan mesafeyi hesapla
        const distSq = this._distanceSquared(node.point, target);
        
        if (distSq < best.distSq) {
            best.distSq = distSq;
            best.point = node.point;
        }

        // Hangi eksende böldüğümüzü belirle
        const axis = node.depth % this.dimensions;
        const key = axis === 0 ? 'x' : 'y';
        const diff = target[key] - node.point[key];

        // Hedef noktaya daha yakın olan dalı önce ara
        const first = diff < 0 ? node.left : node.right;
        const second = diff < 0 ? node.right : node.left;

        // Yakın dalı ara
        this._searchNearest(first, target, best);

        // Bölme düzlemine olan mesafe en iyi mesafeden küçükse
        // diğer dalı da kontrol et (budama / pruning)
        if (diff * diff < best.distSq) {
            this._searchNearest(second, target, best);
        }
    }

    // --- K En Yakın Komşu (KNN) ---

    /**
     * Verilen noktaya en yakın K durağı bulur
     * 
     * Algoritma:
     *   - Sınırlandırılmış sıralı liste (bounded sorted list) kullanır
     *   - K elemandan fazlası olduğunda en uzaktaki atılır
     *   - KD-Tree'nin budama (pruning) özelliği sayesinde
     *     gereksiz dallar atlanır
     * 
     * @param {Object} target - { x, y } hedef nokta
     * @param {number} k - Kaç komşu bulunacağı
     * @returns {Array} En yakın K nokta (mesafe bilgisiyle)
     * @complexity O(K log N) ortalama
     */
    knn(target, k) {
        if (this.root === null || k <= 0) return [];

        // Sonuçları mesafeye göre sıralı tutan liste
        // En uzak eleman sonda (max-heap yerine sıralı dizi)
        const results = [];
        
        this._knnSearch(this.root, target, k, results);

        // Mesafe bilgisiyle birlikte döndür
        return results.map(r => ({
            ...r.point,
            distance: Math.sqrt(r.distSq)
        }));
    }

    _knnSearch(node, target, k, results) {
        if (node === null) return;

        // Mevcut düğüme olan mesafe (karesi)
        const distSq = this._distanceSquared(node.point, target);

        // Listeye ekle veya güncelle
        if (results.length < k) {
            // Henüz K eleman toplanmadı, doğrudan ekle
            results.push({ point: node.point, distSq });
            // Mesafeye göre sırala (küçükten büyüğe)
            results.sort((a, b) => a.distSq - b.distSq);
        } else if (distSq < results[results.length - 1].distSq) {
            // Bu nokta, en uzak komşudan daha yakın
            results[results.length - 1] = { point: node.point, distSq };
            results.sort((a, b) => a.distSq - b.distSq);
        }

        // Hangi eksende böldüğümüzü belirle
        const axis = node.depth % this.dimensions;
        const key = axis === 0 ? 'x' : 'y';
        const diff = target[key] - node.point[key];

        // Hedef noktaya daha yakın olan dalı önce ara
        const first = diff < 0 ? node.left : node.right;
        const second = diff < 0 ? node.right : node.left;

        // Yakın dalı ara
        this._knnSearch(first, target, k, results);

        // Budama: Diğer dalı kontrol etmemiz gerekiyor mu?
        // En uzak komşuya olan mesafe > bölme düzlemine olan mesafe ise evet
        const maxDistSq = results.length < k 
            ? Infinity 
            : results[results.length - 1].distSq;

        if (diff * diff < maxDistSq) {
            this._knnSearch(second, target, k, results);
        }
    }

    // --- Yardımcı Fonksiyonlar ---

    /**
     * İki nokta arasındaki Öklid mesafesinin karesi
     * Kare kök hesaplamaktan kaçınarak performans kazanılır
     * (Karşılaştırma için mesafenin karesi yeterlidir)
     */
    _distanceSquared(a, b) {
        const dx = a.x - b.x;
        const dy = a.y - b.y;
        return dx * dx + dy * dy;
    }

    /**
     * İki nokta arasındaki Öklid mesafesi
     */
    _distance(a, b) {
        return Math.sqrt(this._distanceSquared(a, b));
    }

    /**
     * Ağaçtaki tüm noktaları döndürür (in-order traversal)
     */
    getAllPoints() {
        const points = [];
        this._inOrder(this.root, points);
        return points;
    }

    _inOrder(node, points) {
        if (node === null) return;
        this._inOrder(node.left, points);
        points.push(node.point);
        this._inOrder(node.right, points);
    }

    /**
     * Ağacın derinliğini döndürür
     */
    getHeight() {
        return this._height(this.root);
    }

    _height(node) {
        if (node === null) return 0;
        return 1 + Math.max(
            this._height(node.left), 
            this._height(node.right)
        );
    }

    /**
     * Ağaç istatistiklerini döndürür (debug/analiz için)
     */
    getStats() {
        return {
            nodeCount: this.nodeCount,
            height: this.getHeight(),
            balanceFactor: this.nodeCount > 0 
                ? (Math.log2(this.nodeCount + 1)).toFixed(2) + ' (ideal)'
                : '0',
            actualHeight: this.getHeight()
        };
    }
}

File: src/SmartTransit.Domain/KdTree.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace SmartTransit.Domain
{
    public class KdTreeNode
    {
        public Stop Point { get; }
        public int Depth { get; }
        public KdTreeNode Left { get; set; }
        public KdTreeNode Right { get; set; }

        public KdTreeNode(Stop point, int depth)
        {
            Point = point;
            Depth = depth;
        }
    }

    public class KdTree
    {
        private const int Dimensions = 2;
        private KdTreeNode _root;
        private int _nodeCount;

        // Code Defense gereksinimi: Geliştirici imzası
        public string DeveloperAliTurhan => "Ali Turhan";

        public KdTreeNode Root => _root;
        public int NodeCount => _nodeCount;

        public KdTree(IEnumerable<Stop> points)
        {
            var pointsList = points.ToList();
            _root = BuildBalanced(pointsList, 0);
        }

        private KdTreeNode BuildBalanced(List<Stop> points, int depth)
        {
            if (points.Count == 0) return null;

            int axis = depth % Dimensions;

            if (axis == 0)
            {
                points.Sort((a, b) => a.Latitude.CompareTo(b.Latitude));
            }
            else
            {
                points.Sort((a, b) => a.Longitude.CompareTo(b.Longitude));
            }

            int medianIndex = points.Count / 2;
            var node = new KdTreeNode(points[medianIndex], depth);
            _nodeCount++;

            node.Left = BuildBalanced(points.GetRange(0, medianIndex), depth + 1);
            node.Right = BuildBalanced(points.GetRange(medianIndex + 1, points.Count - (medianIndex + 1)), depth + 1);

            return node;
        }

        public void Insert(Stop point)
        {
            _root = InsertNode(_root, point, 0);
        }

        private KdTreeNode InsertNode(KdTreeNode node, Stop point, int depth)
        {
            if (node == null)
            {
                _nodeCount++;
                return new KdTreeNode(point, depth);
            }

            int axis = depth % Dimensions;
            double nodeCoord = axis == 0 ? node.Point.Latitude : node.Point.Longitude;
            double pointCoord = axis == 0 ? point.Latitude : point.Longitude;

            if (pointCoord < nodeCoord)
            {
                node.Left = InsertNode(node.Left, point, depth + 1);
            }
            else
            {
                node.Right = InsertNode(node.Right, point, depth + 1);
            }

            return node;
        }

        public List<KnnResult> Knn(double targetLat, double targetLon, int k)
        {
            if (_root == null || k <= 0) return new List<KnnResult>();

            var results = new List<KnnResult>();
            KnnSearch(_root, targetLat, targetLon, k, results);

            return results;
        }

        private void KnnSearch(KdTreeNode node, double targetLat, double targetLon, int k, List<KnnResult> results)
        {
            if (node == null) return;

            double distSq = DistanceSquared(node.Point.Latitude, node.Point.Longitude, targetLat, targetLon);

            // Insert or update
            if (results.Count < k)
            {
                results.Add(new KnnResult(node.Point, distSq));
                results.Sort((a, b) => a.DistanceSq.CompareTo(b.DistanceSq));
            }
            else if (distSq < results[results.Count - 1].DistanceSq)
            {
                results[results.Count - 1] = new KnnResult(node.Point, distSq);
                results.Sort((a, b) => a.DistanceSq.CompareTo(b.DistanceSq));
            }

            int axis = node.Depth % Dimensions;
            double nodeVal = axis == 0 ? node.Point.Latitude : node.Point.Longitude;
            double targetVal = axis == 0 ? targetLat : targetLon;
            double diff = targetVal - nodeVal;

            var first = diff < 0 ? node.Left : node.Right;
            var second = diff < 0 ? node.Right : node.Left;

            KnnSearch(first, targetLat, targetLon, k, results);

            double maxDistSq = results.Count < k ? double.PositiveInfinity : results[results.Count - 1].DistanceSq;

            if (diff * diff < maxDistSq)
            {
                KnnSearch(second, targetLat, targetLon, k, results);
            }
        }

        public int GetHeight()
        {
            return CalculateHeight(_root);
        }

        private int CalculateHeight(KdTreeNode node)
        {
            if (node == null) return 0;
            return 1 + Math.Max(CalculateHeight(node.Left), CalculateHeight(node.Right));
        }

        private double DistanceSquared(double lat1, double lon1, double lat2, double lon2)
        {
            double dLat = lat1 - lat2;
            double dLon = lon1 - lon2;
            return dLat * dLat + dLon * dLon;
        }
    }

    public class KnnResult
    {
        public Stop Point { get; }
        public double DistanceSq { get; }
        public double Distance => Math.Sqrt(DistanceSq);

        public KnnResult(Stop point, double distanceSq)
        {
            Point = point;
            DistanceSq = distanceSq;
        }
    }
}
